[all]: Add Support for PowerPC instructions, tests

lukeg101 opened the pull request:
Motivation: we wish to test compilation targeting IBM PowerPC. To do so, we must add new instructions and tests. This PR adds:
- Testing for PPC instructions
- Lis instruction
- HWSYNC fence
- rlwinm and rlwimi instructions
- cmplwi instruction
- addis instruction
- lwa instruction
- nop instruction
- clrldi instruction
- extsw instruction
- fixes blr instruction
- lwax instruction
- Adds a PPC QEMU config
- adds a litmus ppc instruction.h
- emits get_instr in litmus so PPC compiles.

All tests pass in herd, and under litmus QEMU PPC emulation

lukeg101 commented:
This patch is ready for review :)

lukeg101 commented:
Litmus output under QEMU:
```
Sat Sep  2 05:03:48 PM BST 2023
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A000.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A000

{
}
 P0  ;
 nop ;

forall (0:r0=0)
Generated assembler
#START _litmus_P0
	nop
Test A000 Required
Histogram (1 states)
4000000:>0:r0=0;
Ok

Witnesses
Positive: 4000000, Negative: 0
Condition forall (0:r0=0) is validated
Hash=9697a8f62bb095c6209ec10f0eb77308
Observation A000 Always 4000000 0
Time A000 0.90
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A001.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A001

{
}
 P0       ;
 lis r0,1 ;

forall (0:r0=65536)
Generated assembler
#START _litmus_P0
	lis 10,1
Test A001 Required
Histogram (1 states)
4000000:>0:r0=65536;
Ok

Witnesses
Positive: 4000000, Negative: 0
Condition forall (0:r0=65536) is validated
Hash=505b36b29a59339a441f482c0294f18a
Observation A001 Always 4000000 0
Time A001 0.89
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A002.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A002

{
 [P0_r0]=0;
 [P1_r0]=0;
 [x]=0;
 [y]=0;
 uint64_t 0:r1=P0_r0; uint64_t 0:r2=x; uint64_t 0:r5=y;
 uint64_t 1:r1=P1_r0; uint64_t 1:r2=x; uint64_t 1:r5=y;
}
 P0           | P1           ;
 li r4,1      | li r4,1      ;
 stw r4,0(r2) | stw r4,0(r5) ;
 sync         | sync         ;
 lwz r3,0(r5) | lwz r3,0(r2) ;
 stw r3,0(r1) | stw r3,0(r1) ;

exists ([P0_r0]=0 /\ [P1_r0]=0)
Generated assembler
#START _litmus_P0
	li 10,1
	stw 10,0(7)
	sync
	lwz 9,0(6)
	stw 9,0(8)
#START _litmus_P1
	li 10,1
	stw 10,0(6)
	sync
	lwz 9,0(7)
	stw 9,0(8)
Test A002 Allowed
Histogram (3 states)
999954:>[P0_r0]=1; [P1_r0]=0;
999777:>[P0_r0]=0; [P1_r0]=1;
269   :>[P0_r0]=1; [P1_r0]=1;
No

Witnesses
Positive: 0, Negative: 2000000
Condition exists ([P0_r0]=0 /\ [P1_r0]=0) is NOT validated
Hash=d94e9aab81affd834e43c4085405e4c8
Observation A002 Never 0 2000000
Time A002 3.87
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A003.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A003

{
 int 0:r0=0;
}
 P0           ;
 addi r0,r0,1 ;

forall (0:r0=1)
Generated assembler
#START _litmus_P0
	addi 9,9,1
Test A003 Required
Histogram (1 states)
4000000:>0:r0=1;
Ok

Witnesses
Positive: 4000000, Negative: 0
Condition forall (0:r0=1) is validated
Hash=bf27da6f810e02f4302a7c1c8ccaef1c
Observation A003 Always 4000000 0
Time A003 0.89
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A004.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A004

{
}
 P0  ;
 nop ;

forall (0:r0=0)
Generated assembler
#START _litmus_P0
	nop
Test A004 Required
Histogram (1 states)
4000000:>0:r0=0;
Ok

Witnesses
Positive: 4000000, Negative: 0
Condition forall (0:r0=0) is validated
Hash=9697a8f62bb095c6209ec10f0eb77308
Observation A004 Always 4000000 0
Time A004 0.89
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A005.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A005

{
 [x]=1;
 0:r1=x;
}
 P0           ;
 lwz r0,0(r1) ;

forall (0:r0=1)
Generated assembler
#START _litmus_P0
	lwz 10,0(8)
Test A005 Required
Histogram (1 states)
4000000:>0:r0=1;
Ok

Witnesses
Positive: 4000000, Negative: 0
Condition forall (0:r0=1) is validated
Hash=2fe0745fb1662609f725c1740d936d06
Observation A005 Always 4000000 0
Time A005 1.37
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A006.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A006

{
 [x]=1;
 0:r0=0; 0:r1=x;
}
 P0            ;
 lwax r0,r0,r1 ;

forall (0:r0=1)
Generated assembler
#START _litmus_P0
	lwax 10,0,8
Test A006 Required
Histogram (1 states)
4000000:>0:r0=1;
Ok

Witnesses
Positive: 4000000, Negative: 0
Condition forall (0:r0=1) is validated
Hash=b0a3ff4d8b401c000866da026f6f6043
Observation A006 Always 4000000 0
Time A006 1.36
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A007.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A007

{
 [P0_r0]=0;
 [P1_r0]=0;
 [x]=0;
 [y]=0;
 0:r0=x; 0:r1=y; 0:r2=P0_r0;
 1:r0=y; 1:r1=x; 1:r2=P1_r0;
}
 P0              | P1            ;
 sync            | sync          ;
 lwz r9,0(r0)    | lwz r9,0(r0)  ;
 cmpw r9,r9      | cmpw r9,r9    ;
 b   L0x18       | b   L0x6c     ;
 L0x18: isync    | L0x6c: isync  ;
 clrldi r9,r9,32 | extsw r10, r9 ;
 lwsync          | cmpwi r9,1    ;
 sync            | beq  L0x8c    ;
 li r8,1         | L0x7c:        ;
 stw r8,0(r1)    | stw r10,0(r2) ;
 stw r9,0(r2)    | b   exit      ;
                 | L0x8c:        ;
                 | li r8,1       ;
                 | stw r8,0(r1)  ;
                 | b   L0x7c     ;
                 | exit: nop     ;

exists (not ([P0_r0]=1 /\ [P1_r0]=1))
Generated assembler
#START _litmus_P0
	sync
	lwz 10,0(8)
	cmpw 10,10
	b 0f
0:
	isync
	clrldi 10,10,32
	lwsync
	sync
	li 9,1
	stw 9,0(7)
	stw 10,0(6)
#START _litmus_P1
	sync
	lwz 6,0(9)
	cmpw 6,6
	b 0f
0:
	isync
	extsw 5,6
	cmpwi 6,1
	beq 2f
1:
	stw 5,0(8)
	b 3f
2:
	li 7,1
	stw 7,0(10)
	b 1b
3:
	nop
Test A007 Allowed
Histogram (2 states)
1002259*>[P0_r0]=0; [P1_r0]=0;
997741*>[P0_r0]=0; [P1_r0]=1;
Ok

Witnesses
Positive: 2000000, Negative: 0
Condition exists (not ([P0_r0]=1 /\ [P1_r0]=1)) is validated
Hash=81f73d8f08c6c78986131537047c916d
Observation A007 Always 2000000 0
Time A007 4.04
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A008.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A008

{
 0:r1=1;
}
 P0           ;
 extsw r0, r1 ;

forall (0:r0=1)
Generated assembler
#START _litmus_P0
	extsw 10,8
Test A008 Required
Histogram (1 states)
4000000:>0:r0=1;
Ok

Witnesses
Positive: 4000000, Negative: 0
Condition forall (0:r0=1) is validated
Hash=399b3be4b70f7a98559f66704487fbf3
Observation A008 Always 4000000 0
Time A008 0.90
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A009.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A009

{
 int 0:r11=0; int 0:r31=1; 0:r9=1;
}
 P0                ;
 clrldi r11,r31,32 ;
 clrldi r9,r9,32   ;

forall (0:r11=1 /\ 0:r9=1)
Generated assembler
#START _litmus_P0
	clrldi 7,6,32
	clrldi 9,9,32
Test A009 Required
Histogram (1 states)
4000000:>0:r9=1; 0:r11=1;
Ok

Witnesses
Positive: 4000000, Negative: 0
Condition forall (0:r11=1 /\ 0:r9=1) is validated
Hash=9ef5fe48464d5cf42d6e263e7ef93ef6
Observation A009 Always 4000000 0
Time A009 0.93
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A010.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A010

{
 0:r0=2;
}
 P0           ;
 cmplwi r0,2  ;
 beq  L1      ;
 addi r0,r0,1 ;
 L1: nop      ;

forall (0:r0=2)
Generated assembler
#START _litmus_P0
	cmplwi 9,2
	beq 0f
	addi 9,9,1
0:
	nop
Test A010 Required
Histogram (1 states)
4000000:>0:r0=2;
Ok

Witnesses
Positive: 4000000, Negative: 0
Condition forall (0:r0=2) is validated
Hash=7805d6e30c2a9168bf4783dee83e383c
Observation A010 Always 4000000 0
Time A010 0.91
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Results for herd/tests/instructions/PPC/A011.litmus %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
PPC A011

{
}
 P0                   ;
 rlwinm r3,r4,5, 0,31 ;

forall (0:r3=0)
Generated assembler
#START _litmus_P0
	rlwinm 10,8,5,0,31
Test A011 Required
Histogram (1 states)
4000000:>0:r3=0;
Ok

Witnesses
Positive: 4000000, Negative: 0
Condition forall (0:r3=0) is validated
Hash=a2b925fb3895cf50fe0e96ae5027f3b0
Observation A011 Always 4000000 0
Time A011 0.89
Revision 51a39b0e91cd854ca71368e4716d4ea81cbfc78c, version 7.56+03
Command line: ./_build/install/default/bin/litmus7 -mach litmus/libdir/ppc-qemu.cfg -set-libdir litmus/libdir -o out herd/tests/instructions/PPC/@all
Parameters
#define SIZE_OF_TEST 10000
#define NUMBER_OF_RUN 200
#define AVAIL 2
#define STRIDE 1
#define MAX_LOOP 0
/* gcc options: -Wall -std=gnu99 -mbig -pthread */
/* gcc link options: -static */
/* barrier: user */
/* launch: changing */
/* affinity: none */
/* memory: direct */
/* stride: 1 */
/* safer: write */
/* preload: random */
/* speedcheck: no */
/* alloc: dynamic */
/* proc used: 2 */
GCC=powerpc64-linux-gnu-gcc
LITMUSOPTS=
Sat Sep  2 05:04:06 PM BST 2023
```

maranget commented:
Hi @lukeg101. Commit  e59fa8b improves on the initial implementation of "rotate and mask" instructions. I have not looked at the remaining  commits (from `cmllwi` to `lwax`).

But I understand that `blr` is the return instruction and that this instruction is implemented by exiting. This is acceptable under the control of a specific variant. More precisely, when `herd7` is given the option `-variant telechat` (the name is a suggestion), then some instruction such as `blr` can have a specific behaviour that facilitates telechat experiments. Without the variant, instruction have a behaviour as close as possible to machine behaviour: here jump to the address read from some register, or fail when the  instruction is not implemented, which was the case before, if I am not mistaken.

lukeg101 commented:
Hi Luc, thank you for reviewing this! I have added a `telechat` variant to Herd which guards the implementation I added above here. I shall do the same for the other PRs in flight that have similar behaviour

maranget commented:
Hi @lukeg101. At last, some instruction behaviours fixed (in particular signed vs. zero extension), commits squashed and final `rebase master` performed . This PR is candidate for being merged tomorrow. This leaves a little time for a last minute check.

Thanks

maranget commented:
Hi @lukeg101, merged thanks.

lukeg101 commented:
thank you very much, Luc!
